### PR TITLE
Fix idempotency for Unix permissions in zip files.

### LINKED ISSUE #22254: Unarchive modules always reports 'changed' with non-default permissions on items in zip files.  (breaks idempotence)
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME
unarchive module

##### ANSIBLE VERSION

```
ansible 2.2.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### CONFIGURATION
Defaults

##### OS / ENVIRONMENT
Running on OEL 7.3 (RHEL equivalent) 

##### SUMMARY
Zip files containing items with non-default permissions are always reported as changed, every run.  Totally breaking idempotence.

##### STEPS TO REPRODUCE

```
### Create two files, and change the permissions on one of them.
$ touch test-777 test-default
$ chmod 777 test-777

### Put both files in seperate zip files
$ zip test-777.zip test-777
  adding: test-777 (stored 0%)
$ zip test-default.zip test-default
  adding: test-default (stored 0%)

#### Make a playbook that unarchives those.
$ cat ../test.yaml
---
- hosts: localhost

  tasks:

  - name: 777 zip
    unarchive:
      src=test-777.zip
      dest=/


  - name: Default zip
    unarchive:
      src=test-default.zip
      dest=/

###  First run should report 'changed' for both tasks.
$ ansible-playbook ../test.yaml 
 [WARNING]: provided hosts list is empty, only localhost is available

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [777 zip] *****************************************************************
changed: [localhost]

TASK [Default zip] *************************************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=2    unreachable=0    failed=0   

### Verify the files look proper
$ ls -l /test-777 test-777 /test-default test-default
-rwxrwxrwx. 1 root root 0 Mar  3 11:13 /test-777
-rwxrwxrwx. 1 root root 0 Mar  3 11:13 test-777
-rw-r--r--. 1 root root 0 Mar  3 11:10 /test-default
-rw-r--r--. 1 root root 0 Mar  3 11:10 test-default


### Second run should report no changes.
$ ansible-playbook ../test.yaml --diff
 [WARNING]: provided hosts list is empty, only localhost is available

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [777 zip] *****************************************************************
.f...p...?? test-777/
changed: [localhost]

TASK [Default zip] *************************************************************
ok: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=3    **changed=1**    unreachable=0    failed=0   
```

##### EXPECTED RESULTS
The second ansible run should not make/report any changes.  The files already exists properly on the destination, due to the first ansible run

##### ACTUAL RESULTS

The second ansible run reports a change on the file with non-default permissions.

```
$ ansible-playbook test.yaml --diff
 [WARNING]: provided hosts list is empty, only localhost is available

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [777 zip] *****************************************************************
.f...p...?? test-777
changed: [localhost]

TASK [Default zip] *************************************************************
ok: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=3    **changed=1**    unreachable=0    failed=0   

```

### PR DESCRIPTION
**Pull request by chriskarel**
This fix prevents the unarchive module from reporting 'changed' when a zipfile contains items with Unix permissions that differ from the system default.

##### SUMMARY

Fixes #22254
Currently, unarchive will always report changed on any items with high non-default Unix permissions.  This is because the module always subtracts the umask from from any permission set it is evaluating, so 'high' permissions sets, or high umask values, will result in the check always failing.  

Example:  A Umask of 022 will result in files with user or group permissions of '6' or '7' (eg: 706, 770) always being marked as changing, because the existing permissions differ when the UMask is applied.  (The examples are checked against 704 and 750)

This PR simply zeros out the UMask value for files with Unix style permissions.  This results in the comparison being done against the intended permissions set.  My testing indicates it preserves idempotence for archives containing files of Unix permissions, Windows permissions, and both.

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

Unarchive module

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Sep  5 2016, 02:30:38) [GCC 4.8.5 20150623 (Red Hat 4.8.5-9)]

```

##### ADDITIONAL INFORMATION

--For Dennis.  I love you little buddy.

```

```

### CODE REVIEW

**Comment by dagwieers** [comment]
I approve of your change, but we also need integration tests for this eventually.

shipit

**Comment by chriskarel** [comment]
OK, I wrote some additional CI tests for this issue, and in doing so:

A) Accidentally committed them into this PR.  I think I've reverted them, but it obviously made things messy.
B) Realized that my initial fix caused regressions under some mixed-content zip files.  I've made additional changes to fix this issue.

Let me know if I need to recreate this PR, or if there are any further steps to get it merged.

**Comment by dagwieers** [comment]
The integration tests should be part of this PR if you have them. As they prove that the changes work the way they are expected to work. Which makes the review easier, because we can also assess the expectations.

**Comment by mattclay** [comment]
CI failure in integration tests on multiple platforms. Here's one example:

https://app.shippable.com/github/ansible/ansible/runs/21886/26/tests

**Comment by chriskarel** [comment]
Good grief, what a rabbit hole I've gone down on this one.

The initial issue of #22254 is fixed.

However, BSD's unzip acts completely opposite to all other versions when it comes to permissions and UMask -- it always applies the umask value, overriding what is specified in the archive.  So there's now a test for a BSD system, with a change in behavior.

In writing the Unit tests, I discovered that the versions of Unzip used by Fedora and Ubuntu have a bug that impacts the timestamps on extracted files.  To work around that, I've had to apply specific timestamps to test files.  (That doesn't trigger the bug, no idea why)  The extraneous touch can be removed when the underlying bugs are fixed.  ([Fedora](https://bugzilla.redhat.com/show_bug.cgi?id=1451953), [Ubuntu](https://bugs.launchpad.net/ubuntu/+source/unzip/+bug/1691636))

The idempotence test also exposed the fact that BSD's unzip doesn't support listing timestamps in zip archives.  This resulted in unarchive always unzipping on that OS.  So I've configured the unarchive module to use zipinfo instead.  And the CI script was modified to make sure the unzip package is installed (Which wasn't, on BSD).

This PR is now obviously a cavalcade of commit nightmares.  I'm more than happy to recreate this if anyone objects to the mess.  Otherwise, it will just stand as a testament to how much of an iceberg this bug became.  But, if there's anything else I should do, just let me know.

**Comment by chriskarel** [comment]
Hey, @dagwieers, what more do I need to do to get this approved?  Are we waiting on @pileofrogs as a second `shipit` approver?

**Comment by dagwieers** [comment]
Yes, or one of the core members. You may want to bring this up on one of the core meetings...

File: lib/ansible/modules/files/unarchive.py [hunk 1]
```diff
@@ -391,13 +395,24 @@ def is_unarchived(self):
...
+
             if len(permstr) == 6:
                 if path[-1] == '/':
                     permstr = 'rwxrwxrwx'
                 elif permstr == 'rwx---':
                     permstr = 'rwxrwxrwx'
                 else:
                     permstr = 'rw-rw-rw-'
+                file_umask=umask
+            elif 'bsd' in 'bsd' in systemtype.lower():
```
Comment (sivel) [inline lib/ansible/modules/files/unarchive.py]: This looks like a typo.  I think you just want `elif 'bsd' in systemtype.lower()`

**Review by sivel — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/files/unarchive.py (hunk 1 above)
Comment (chriskarel) [reply lib/ansible/modules/files/unarchive.py]: Sure is.  Good catch.  Fixed.

**Comment by dagwieers** [comment]
@chriskarel Welcome to my world :-/ I actually wanted to ditch FreeBSD at one point because of how hard it was to keep supporting it. But the main mess comes from the fact we are depending on an external tool and interface via its stdout and stderr and that's very prone to errors.

In fact I think we should work towards a new pure-python **uncompress** module. It's going to be much easier to support.

PS The fact that FreeBSD would be overwriting files every time is not that bad a behaviour, it used to be the normal behaviour for all platforms a year ago ;-)
PS2 I have a big PR ready that cleans up the module and makes it a lot easier to debug when people report issues. Unfortunately, the rewrite failed to work on FreeBSD only (because of its awkwardness compared to all other platforms). #19015

**Comment by abadger** [comment]
I've rebased this PR for you.  Waiting on shippable tests.  If that passes I'll squash and merge.

**Comment by abadger** [comment]
This passed -- One of the centos builds failed due to the yum repository being unavailable and then succeeded on the retry.

**Comment by abadger** [comment]
Okay, merged to devel for the 2.4.0 release.   I'm not sure about whether it should be backported for a possible 2.3.3 release or not.  @dagwieers if you think it is safe and should be backported, let me know on IRC and I'll cherry-pick it.

**Pull request merged**